spec/features/api_particulier/prolong_token_build_spec.rb

```diff
@@ -0,0 +1,185 @@
+RSpec.describe 'follows the prolong token wizard', app: :api_particulier do
+  subject(:go_to_prolong_token_start) do
+    visit api_particulier_token_prolong_start_path(token_id: token.id)
+  end
+
+  let!(:authenticated_user) { create(:user) }
+  let!(:non_authenticated_user) { create(:user, :demandeur) }
+  let!(:exp) { 88.days.from_now.to_i }
+  let!(:api) { 'entreprise' }
+
+  let!(:authorization_request) do
+    create(
+      :authorization_request,
+      :with_demandeur,
+      demandeur: authenticated_user,
+      api:,
+      status: 'validated'
+    )
+  end
+
+  let!(:token) do
+    create(:token, authorization_request:, exp:)
+  end
+
+  describe 'when user is not authenticated' do
+    it 'redirects to the login' do
+      go_to_prolong_token_start
+      expect(page).to have_current_path(api_particulier_login_path, ignore_query: true)
+    end
+  end
+
+  describe 'when user is authenticated' do
+    before do
+      login_as(authenticated_user)
+    end
+
+    describe 'when contact_technique or contact_metier is not present' do
+      it 'does not allow to keep contacts' do
+        go_to_prolong_token_start
+
+        choose('prolong_token_wizard_owner_still_in_charge')
+        click_button('next_prolong_token_wizard')
+        expect(current_url).to include('project_purpose')
+
+        choose('prolong_token_wizard_project_purpose_true')
+        click_button('next_prolong_token_wizard')
+        expect(current_url).to include('contacts')
+
+        expect(page).to have_no_css('#prolong_token_wizard_contact_metier_true')
+        expect(page).to have_no_css('#prolong_token_wizard_contact_technique_true')
+
+        expect(page).to have_css('#prolong_token_wizard_contact_metier_false')
+        expect(page).to have_css('#prolong_token_wizard_contact_technique_false')
+      end
+    end
+
+    describe 'when on happy path with api entreprise' do
+      let!(:authenticated_user) { create(:user, :demandeur, :contact_metier, :contact_technique) }
+
+      let!(:authorization_request) do
+        create(
+          :authorization_request,
+          :with_demandeur,
+          :with_contact_metier,
+          :with_contact_technique,
+          demandeur: authenticated_user,
+          contact_metier: authenticated_user,
+          contact_technique: authenticated_user,
+          api:,
+          status: 'validated'
+        )
+      end
+
+      it 'is prolonged' do
```

Comment:
```suggestion
      it 'prolongs the token' do
```

c'est plus clair imo

Reply:
Bon en vrai je refactoriserai ces 2 it en 2 blocks `context/it` pour que ça soit plus clair

```ruby
describe 'fill the wizard' do
  context 'when we valid all questions' do
    it 'prolongs the token'
  end
  
  context 'when we check that at least one field has to change' do
    it 'requires an update on DataPass'
  end
end
```

Comme ça ici tu peux faire un mega subject `fill_wizard` dans le describe:

```ruby
subject(:fill_wizard) do
  go_to_prolong_token_start
  
  # `in_charge_option` est défini à true/false en fonction du context
  choose(in_charge_option)
  click_button('next_prolong_token_wizard')
  # expect(current_url).to include('project_purpose')
  
  choose('prolong_token_wizard_project_purpose_true')
  click_button('next_prolong_token_wizard')
  # expect(current_url).to include('contacts')
  
  # expect(page).to have_css('#prolong_token_wizard_contact_metier_true')
  # `prolong_contact_metier_option` est défini à true/false en fonction du context
  choose(prolong_contact_metier_option)
  # expect(page).to have_css('#prolong_token_wizard_contact_technique_true')
  # `prolong_contact_technique_option` est défini à true/false en fonction du context
  choose(prolong_contact_technique_option)
  click_button('next_prolong_token_wizard')

  # expect(current_url).to include('summary')
  
  click_link('finished_prolong_token_wizard')
  # expect(current_url).to include('finished')
end
```

J'ai commenté tous les `expect` car ils ne servent pas à mon sens ici (je suppose que tu as fait ça pour t'aider à coder).

Reply:
Et d'ailleurs c'est un refactor logique, tu te répètes 4 fois, tu pourrais même remonté le subject plus haut en faisant une disjonction sur le type d'app (entreprise ou particulier)

Reply:
Ok ca me semble pas mal en effet.